### DISCUSSION TITLE
Logo Interpreter

### DISCUSSION THREAD

**Discussion by CaptainBoggle** [badicsalex/peginator#28] [Show and tell]
Hi! I just wanted to say a huge thank you, because this library has been an absolute dream to work with as a rust beginner, especially compared to other more popular parser generator or combinator libraries that I have tried. Using peginator, I wrote an interpreter for an extended version of the Logo language. Unfortunately I can't share my work here as it was completed for a university assignment, but if you are interested I would be happy to share it privately. If I have time, I might also use this for a personal project of mine. If there is any way I could contribute to peginator, please let me know, as I would love to repay the project somehow!

**Comment by badicsalex** [comment]
I appreciate the feedback, it means a lot :)
Especially since Logo was the first programming language I learned, so it has a special place in my heart.

The fact you had fun is enough contribution, no need to "repay" anything. Maybe if you had some ideas on how the library could have been better while doing your assignment, you could share those.

**Reply by CaptainBoggle** [reply]
What an awesome coincidence that Logo was your first language! As for some ideas I had:

First, I would love if you could add a fully custom ParseErrorSpecifics variant that simply outputs a given string with no additional processing, as I ended up using your error handling for a few runtime checks to keep the error style consistent, but had to use the "extern rule" specific because it made the most sense out of the options, despite not being totally accurate. Also, maybe consider adding a mechanism for people to override the generated error messages in certain situations where they could be confusing. Finally for errors, possibly support the use of a span/range rather than single byte position to describe error locations. Thinking about it more, maybe add a crate feature that replaces the simple error reporting with some miette or ariadne integration so people can decide which better suit their use case, as I understand that lots of applications won't want to use locational errors outside of parsing. 

Second, clippy/rust-analyzer find some issues with the generated code, like redundant usage of super:: in some places. I also had some clippy::enum_variant_names warnings, but they came from the way I structured my EBNF, not peginator itself.

**Reply by badicsalex** [reply]
Ah, these are great ideas to improve error reporting, thanks. I'll probably create a bunch of github issues to have them in a more structured, visible place.
I have never been too satisfied with the error reporting anyway, but [my own use-case](https://github.com/badicsalex/hun_law_rs) didn't actually need it, so I didn't have the motivation to fix it :)
